### PR TITLE
Do batch inserts in create_rollup_tables

### PR DESCRIPTION
**Pull request by islemaster**
While trying to track down a performance issue last night, we noticed that the `create_rollup_tables` task that runs at 1am PST every night truncates three tables and then inserts more than 30,000 rows using individual queries.  That's a lot of queries to run from one process in a few seconds.

I've modified the script to do inserts in batches of 2000 rows, so now it should generate dozens of queries instead of thousands.  I've also removed the transactions, which don't provide any protection against the `TRUNCATE` operations anyway.

I think there's a larger discussion that needs to happen soon:
- The output of this task is only used for analytics - why does it run on production-daemon, and why does it write to our production database?

## Testing story

Re-run locally with and without transactions, with different batch sizes, spot-checking that the same rows (and same number of rows) get inserted into the target tables.

# Reviewer Checklist:

- [ ] Tests provide adequate coverage
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: bin/cron/create_rollup_tables [hunk 1]
```diff
@@ -143,92 +145,96 @@ def main
...
+        created_at,
+        updated_at,
+        level_id,
+        answer_number,
+        answer_text,
+        correct
       )
-    end
+      VALUES
+      #{batch.map {|row| contained_level_answers_row(row)}.join(",\n")}
```
Comment (bencodeorg) [inline bin/cron/create_rollup_tables:182]: Slick!

**Comment by bencodeorg** [comment]
Just to comment on the question posed here about why this is happening in production, my impression is there are benefits to having access to all of our Ruby application logic in building these tables. 

I could see reading from production, and writing directly to Redshift (we already have support for [connecting to Redshift](https://github.com/code-dot-org/code-dot-org/blob/staging/lib/cdo/redshift.rb). Would that alleviate your concerns @islemaster, or do you think even reads should be moved elsewhere?

Worth tracking that the [script](https://github.com/code-dot-org/code-dot-org/blob/staging/bin/cron/process_jotform_data) that processes Jotform data (and has caused database problems in the past and as a result is currently disabled) is heavily inspired by the script in this PR.

**Comment by islemaster** [comment]
Ah, I see what you're saying about the application logic.  I do think writing directly to Redshift would be an improvement here - I'd like to keep moving things out of the production database that aren't actually used by the application.

**Comment by bencodeorg** [comment]
Looks good to me -- shoot me a message when you merge this so I can confirm that everything looks reasonable once it runs via cron?

**Comment by islemaster** [comment]
Hey @bencodeorg , you mentioned the `process_jotform_data` script but it looks like you disabled it in late October: https://github.com/code-dot-org/code-dot-org/pull/31399

Is that something we intend to re-enable, or should we just remove that script?

**Comment by sureshc** [comment]
Does inserting in batches of 1000 impact performance by keeping a transaction open for 1000 inserts?  Could we just insert each row in its own transaction?

**Comment by bencodeorg** [comment]
It needs to be turned back on, but I didn't want to do anything until we could improve performance. Chatted with @hacodeorg a bit about how to do it, but I haven't had time to get into it.

**Comment by islemaster** [comment]
It's significantly faster (and better for overall performance) to insert many rows at once, because it allows MySQL to deduplicate a lot of work. From [8.2.4.1 Optimizing INSERT Statements](https://dev.mysql.com/doc/refman/5.7/en/insert-optimization.html):

> To optimize insert speed, combine many small operations into a single large operation. Ideally, you make a single connection, send the data for many new rows at once, and delay all index updates and consistency checking until the very end.

The docs go on to describe the work involved in processing an individual query with rough sizes suggesting it may be about 10x faster to insert 1000 rows in a single query than to do so individually.

**Comment by islemaster** [comment]
But your question was about _overall_ performance, not just this particular query.  Assuming the simple 10x speed improvement described above, we are still choosing (for example) between one thousand 10ms queries (total 10 seconds), or one 1000ms query.  Is it better to do the 10ms queries so we avoid the long 1000ms query interfering with other database operations?

[Here's an "aggregated" answer to that question.](https://stackoverflow.com/a/37382695) The gist seems to be: "As a rule of thumb, the less queries the better.  But it depends."

In this particular case if we assume that the queries will create a write-lock on the destination table, I still think the large query is a better call because these are reporting tables that nothing else in our system is writing to, and nothing at the application layer should be reading.

**Comment by islemaster** [comment]
What do we think, team? Worth proposing as fixing a HOC scale risk, or should I point this PR at staging-next?

**Comment by sureshc** [comment]
I think this change is safe to deploy before Hour of Code.  The risk of NOT optimizing the performance of this query is low, since it runs at 1AM PST.

**Review by sureshc — APPROVED** [review]

**Review by bencodeorg — APPROVED** [review]
I didn't catch whether the index you guys added to `pd_attendances` fully solved the blip you saw, or if this is still causing some visible impact? Seems like a good improvement, but if the main problem was `pd_attendances` not sure if it's worth "going to bat" for this change at the PR triage or just wait until after HoC.

**Comment by islemaster** [comment]
The [`pd_attendances` indexes](https://github.com/code-dot-org/code-dot-org/pull/32083) did fix most of our nightly write latency issues. Here's a one week view, and you can see how the nightly spike is _nearly_ eliminated about five days ago.

image

Adjusting the scale with a view of the last three days, we can still see much smaller spikes in write latency, but none of them seem to correspond to this 1am job.

image

That said, I'm still not satisfied with the explanation that the midnight account purge task is responsible for the 1am spike that started this whole investigation.  I won't be too concerned if we don't ship this change for HOC, but since it's quite low-risk I'll bring it to triage for consideration.

**Pull request merged**